### PR TITLE
Built-in atomic functions added

### PR DESCRIPTION
**Pull request by nidin**
Built-in atomic functions added for `SharedArrayBuffer`.

# List of functions
```typescript
Atomic.load<T>(offset: usize, constantOffset?: usize): T
Atomic.store<T>(offset: usize, value: void, constantOffset?: usize): void
Atomic.add<T>(ptr: usize, value: T, constantOffset?: usize): T
Atomic.sub<T>(ptr: usize, value: T, constantOffset?: usize): T
Atomic.and<T>(ptr: usize, value: T, constantOffset?: usize): T
Atomic.or<T>(ptr: usize, value: T, constantOffset?: usize): T
Atomic.xor<T>(ptr: usize, value: T, constantOffset?: usize): T
Atomic.xchg<T>(ptr: usize, value: T, constantOffset?: usize): T
Atomic.cmpxchg<T>(ptr: usize, expected:T, replacement: T, constantOffset?: usize): T
Atomic.wait<T>(ptr: usize, expected:T, timeout:i64): i32
Atomic.notify<T>(ptr: usize, count: u32): u32
```

### CODE REVIEW

File: src/builtins.ts [hunk 1]
```diff
@@ -2844,7 +3333,18 @@ function deferASM(
   valueType: Type,
   reportNode: Node
 ): ExpressionRef {
-  var prototype = assert(compiler.program.elementsLookup.get(name));
```
Comment (dcodeIO) [inline src/builtins.ts]: `elementsLookup` should also contain namespace members here, like both `Atomic` and `Atomic.load` etc. Is there a bug?

File: src/builtins.ts [hunk 2]
```diff
@@ -707,13 +709,13 @@ export function compileCall(
...
               ),
-              module.createGetLocal(tempLocalIndex1, NativeType.I64)
-            ),
-            module.createGetLocal(tempLocalIndex2, NativeType.I64)
-          );
+              module.createGetLocal(tempLocalIndex2, NativeType.I64)
+            );
 
-          currentFunction.freeTempLocal(tempLocal1);
+            currentFunction.freeTempLocal(tempLocal1);
```
Comment (dcodeIO) [inline src/builtins.ts]: nit: indentation seems to be off

File: src/builtins.ts (hunk 1 above)
Comment (nidin) [reply src/builtins.ts]: I will cross check. `elementsLookup` was not in `elementsLookup` but it was a few months ago.

Comment (nidin) [reply src/builtins.ts]: For some reason namespace members are `undefined` in `elementsLookup`
```typescript
compiler.program.elementsLookup.get("Atomic.load")
```
return `undefined`

---------------
```typescript
compiler.program.elementsLookup.get("Atomic")
```
return 
```typescript
MapIterator {
  'load',
  'store',
  'add',
  'sub',
  'and',
  'or',
  'xor',
  'xchg',
  'cmpxchg',
  'wait',
  'notify' }
```

File: src/builtins.ts [hunk 3]
```diff
@@ -1668,6 +1670,416 @@ export function compileCall(
...
+        operands[0],
+        compiler.options.usizeType,
+        ConversionKind.IMPLICIT,
+        WrapMode.NONE
+      );
+      let offset = operands.length == 2 ? evaluateConstantOffset(compiler, operands[1]) : 0; // reports
+      if (offset < 0) { // reported in evaluateConstantOffset
+        return module.createUnreachable();
+      }
+      compiler.currentType = typeArguments[0];
```
Comment (MaxGraey) [inline src/builtins.ts]: It seems `typeArguments[0]` used 5 times so it will be great cache it to variable

File: src/builtins.ts [hunk 4]
```diff
@@ -1668,6 +1670,416 @@ export function compileCall(
...
+        offset
+      );
+    }
+    case "Atomic.store": { // Atomic.store<T!>(offset: usize, value: *, constantOffset?: usize) -> void
+      compiler.currentType = Type.void;
+      if (operands.length < 2 || operands.length > 3) {
+        if (!(typeArguments && typeArguments.length == 1)) {
+          compiler.error(
+            DiagnosticCode.Expected_0_type_arguments_but_got_1,
+            reportNode.range, "1", typeArguments ? typeArguments.length.toString(10) : "0"
```
Comment (MaxGraey) [inline src/builtins.ts]: `.toString(10)` is unnecessary. Just use `toString()`

File: src/builtins.ts (hunk 3 above)
Comment (nidin) [reply src/builtins.ts]: I will change it.

File: src/builtins.ts (hunk 4 above)
Comment (nidin) [reply src/builtins.ts]: This too.

**Comment by MaxGraey** [comment]
btw you could try turn on threads for latest node via `node --experimental-wasm-threads` and write some tests.

**Comment by willemneal** [comment]
How do threads work with node?  In the Browser there are web workers, but I thought node was single threaded.

**Comment by MaxGraey** [comment]
Mostly the same for now: https://nodejs.org/api/worker_threads.html

**Comment by nidin** [comment]
@MaxGraey  This PR only includes built-in functions to make review easy. PR #278 Contains everything but it's too much for a single review. I am actually breaking the PR #278 into small parts.

**Comment by MaxGraey** [comment]
I see. Totally makes sense. But also will great have small test cases of functionality present in this PR. Or add it in separate PR

**Comment by nidin** [comment]
Atomic operations won't compile without shared memory support in compiler. It will come in next PR. followed by test cases.

**Comment by MaxGraey** [comment]
Got it!

**Comment by dcodeIO** [comment]
Great, thanks! :)

**Pull request merged**